telemetry/telemetry_linux.go

```diff
@@ -0,0 +1,273 @@
+	DiskFree  uint64
+}
+
+const (
+	TelemetryFile = "/var/run/AzureCNITelemetry.json"
+	MB            = 1048576
+	KB            = 1024
+)
+
+// Read file line by line and return array of lines.
+func ReadFileByLines(filename string) ([]string, error) {
+	var (
+		lineStrArr []string
+	)
+
+	f, err := os.Open(filename)
+	if err != nil {
+		return nil, fmt.Errorf("Error opening %s file error %v", filename, err)
+	}
+
+	r := bufio.NewReader(f)
+
+	for {
+		lineStr, err := r.ReadString('\n')
+		if err != nil {
+			if err != io.EOF {
+				return nil, fmt.Errorf("Error reading %s file error %v", filename, err)
+			}
+			break
+		}
+		lineStrArr = append(lineStrArr, lineStr)
+	}
+
+	err = f.Close()
+	if err != nil {
+		return nil, fmt.Errorf("Error closing %s file error %v", filename, err)
+	}
+
+	return lineStrArr, nil
+}
+
+// This function retrieves VMs memory usage.
+func getMemInfo() (*MemInfo, error) {
+	var (
+		memArray []uint64
+	)
+
+	linesArr, err := ReadFileByLines("/proc/meminfo")
+	if err != nil {
+		return nil, err
+	}
+
+	for _, line := range linesArr {
+		fieldArray := strings.Split(line, ":")
+		fieldArray[0] = strings.TrimSpace(fieldArray[0])
+		if strings.Compare(fieldArray[0], "MemTotal") == 0 {
+			memTotalStr := strings.TrimLeft(fieldArray[1], " ")
+			memTotalArr := strings.Split(memTotalStr, " ")
+			memTotal, err := strconv.ParseUint(memTotalArr[0], 0, 64)
+			if err != nil {
+				return nil, fmt.Errorf("Failed in atoi conversion of memtotal %v", err)
+			}
+			memTotal = memTotal / KB
+			memArray = append(memArray, memTotal)
+		} else if strings.Compare(fieldArray[0], "MemFree") == 0 {
+			memFreeStr := strings.TrimLeft(fieldArray[1], " ")
+			memFreeArr := strings.Split(memFreeStr, " ")
+			memFree, err := strconv.ParseUint(memFreeArr[0], 0, 64)
+			if err != nil {
+				return nil, fmt.Errorf("Failed in atoi conversion of memtotal %v", err)
+			}
+
+			memFree = memFree / KB
+			memArray = append(memArray, memFree)
+		}
+	}
+
+	memInfo := &MemInfo{MemTotal: memArray[0], MemFree: memArray[1]}
+
+	return memInfo, nil
+}
+
+// This function retrieves VMs disk usage.
+func getDiskInfo(path string) (*DiskInfo, error) {
+	fs := syscall.Statfs_t{}
+	err := syscall.Statfs(path, &fs)
+	if err != nil {
+		return nil, fmt.Errorf("Statfs call failed with error %v", err)
+	}
+
+	total := fs.Blocks * uint64(fs.Bsize) / MB
+	free := fs.Bfree * uint64(fs.Bsize) / MB
+
+	diskInfo := &DiskInfo{DiskTotal: total, DiskFree: free}
+	return diskInfo, nil
+}
+
+// This function  creates a report with system details(memory, disk, cpu).
+func GetSystemDetails() (*SystemInfo, error) {
+	var cpuCount int = 0
+
+	linesArr, err := ReadFileByLines("/proc/cpuinfo")
+	if err != nil {
+		return nil, err
+	}
+
+	for _, line := range linesArr {
+		fieldArray := strings.Split(line, ":")
+		fieldArray[0] = strings.TrimSpace(fieldArray[0])
+		if strings.Compare(fieldArray[0], "processor") == 0 {
+			cpuCount += 1
+		}
+	}
+
+	memInfo, err := getMemInfo()
```

Comment:
if it returns error, lets continue to get next info. log error inside the specific error details that we will add.